source/expressions.tex

```diff
@@ -2610,45 +2610,43 @@
-
-\item If the \grammarterm{return-type-requirement} is present, then:
-
-\begin{itemize}
-\item Substitution of template arguments (if any)
-into the \grammarterm{return-type-requirement} is performed.
-
-\item If the \grammarterm{return-type-requirement} is a
-\grammarterm{trailing-return-type},
-%%% FIXME: is -> shall be
-\tcode{E} is implicitly convertible to
-the type named by the \grammarterm{trailing-return-type}.
-If conversion fails, the enclosing \grammarterm{requires-expression}
-is \tcode{false}.
-
-\item If the \grammarterm{return-type-requirement}
-starts with a \grammarterm{constrained-parameter}\iref{temp.param},
-the \grammarterm{expression} is deduced against
-an invented function template \tcode{F}
-using the rules in \ref{temp.deduct.call}.
-\tcode{F} is a \tcode{void} function template
-with a single type template parameter \tcode{T}
-declared with the \grammarterm{constrained-parameter}.
-A \grammarterm{cv-qualifier-seq} \cv{} is formed
-as the union of \tcode{const} and \tcode{volatile} specifiers
-around the \grammarterm{constrained-parameter}.
-\tcode{F} has a single parameter
-whose \grammarterm{type-specifier} is \cv{}~\tcode{T}
-followed by the \grammarterm{abstract-declarator}.
-%%% FIXME: Remove this; if deduction fails, the construct should
-%%% be ill-formed.
-If deduction fails,
-the enclosing \grammarterm{requires-expression} is \tcode{false}.
-\end{itemize}
+\item
+  Substitution of template arguments (if any)
+  into the \grammarterm{expression} is performed.
+\item
+  If the \tcode{noexcept} specifier is present,
+  \tcode{E} shall not be a potentially-throwing expression\iref{except.spec}.
+\item
+  If the \grammarterm{return-type-requirement} is present, then:
+  \begin{itemize}
+  \item
+    Substitution of template arguments (if any)
+    into the \grammarterm{return-type-requirement} is performed.
+  \item
+    If the \grammarterm{return-type-requirement} is a
+    \grammarterm{trailing-return-type}\iref{dcl.decl},
+    %%% FIXME: is -> shall be
+    \tcode{E} is implicitly convertible to
+    the type named by the \grammarterm{trailing-return-type}.
+    If conversion fails, the enclosing \grammarterm{requires-expression}
+    is \tcode{false}.
+  \item
+    If the \grammarterm{return-type-requirement}
+    is of the form \tcode{->} \grammarterm{type-constraint}, then
+    the contextually-determined type being constrained
+    is \tcode{decltype((E))}.
+    The immediately-declared constraint\iref(temp) of \tcode{decltype((E))}
+    shall be satisfied.
+    \begin{note}
+    Thus, constraints of the form
+    \tcode{\{ E \} -> Concept;} or of the form
+    \tcode{\{ E \} -> Concept<>;} are equivalent to
+    \tcode{E; requires Concept<decltype((E))>;}
+    while a constraint of the form
+    \tcode{\{ E \} -> Concept<A$_1$, A$_2$, . . . , A$_n$>;} is equivalent to
+    \tcode{E; requires Concept<decltype((E)), A$_1$, A$_2$, ..., A$_n$>;}.
+    \end{note}
```

Comment:
This note is overall a little hard to read due to the frequent switching to and from code font and the `E; requires ...;` construction appearing to be two separate clauses. Perhaps a different presentation would help -- maybe converting the `tcode` here into `codeblock`s, or more drastically something like:

> [*Example*: Given a concept named `X`,
> ```
> requires {
>   { E1 } -> X;
>   { E2 } -> X<A1, A2, ..., An>;
> }
> ```
> is equivalent to
> ```
> requires {
>   E1; requires X<decltype((E1))>;
>   E2; requires X<decltype((E2)), A1, A2, ..., An>;
> };
> ```
> -- *end example*]

@W-E-Brown What do you think?